src/Data/Singletons/Promote.hs

```diff
@@ -391,15 +436,25 @@ promoteMethod m_subst sigs_map (meth_name, meth_rhs) = do
     lookup_meth_ty :: PrM ([DKind], DKind)
     lookup_meth_ty = case Map.lookup meth_name sigs_map of
       Nothing -> do
-        mb_info <- dsReify proName
-        case mb_info of
-          Just (DTyConI (DOpenTypeFamilyD (DTypeFamilyHead _ tvbs mb_res_ki _)) _)
-            -> let arg_kis = map (default_to_star . extractTvbKind) tvbs
-                   res_ki  = default_to_star (resultSigToMaybeKind mb_res_ki)
-               in return (arg_kis, res_ki)
-          _ -> fail $ "Cannot find type annotation for " ++ show proName
+        let mk_meth_ty = extract_meth_ty (dsReify proName)
+                     <|> extract_meth_ty (dsReifyTypeNameInfo proName)
+                     -- See Note [Using dsReifyTypeNameInfo when promoting instances]
```

Comment:
Hold the phone. We definitely don't want to go with the `mkTypeNameWithLocals` approach - at least not while [GHC Trac #13968](https://ghc.haskell.org/trac/ghc/ticket/13968) is around. That's because TH will try to splice in exact `Name`s for promoted things (instead of `Name`s that were created from just `mkName`). When combined with GHCi, this can lead to horrifying results like this that I just observed:

```
λ> ; $(singletons [d| f :: Bool; f = True |])
<interactive>:4:5-41: Splicing declarations
    singletons
      [d| f_a206O :: Bool
          f_a206O = True |]
  ======>
    f_a207e :: Bool
    f_a207e = True
    type FSym0 = interactive:Ghci1.F
    type family interactive:Ghci1.F :: Bool where
      = TrueSym0
    sF :: Sing (FSym0 :: Bool)
    sF = STrue
λ> ; $(singletons [d| f :: Bool; f = True |])
<interactive>:5:5-41: Splicing declarations
    singletons
      [d| f_a207Z :: Bool
          f_a207Z = True |]
  ======>
    f_a208p :: Bool
    f_a208p = True
    type FSym0 = F
    type family F :: Bool where
      = TrueSym0
    sF :: Sing (FSym0 :: Bool)
    sF = STrue

<interactive>:5:5: error: Illegal binding of built-in syntax: F
```

I'll revert the `mkTypeNameWithLocals` changes here, since the previous iteration of this pull request didn't suffer from this problem.